docs/src/extend/custom-rules.md

```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+                type: "object",
+                properties: {
+                    exceptRange: { type: "boolean" }
                 },
-                "additionalProperties": false
+                additionalProperties: false
             }
         ]
-    },
+    }
 };
 ```
 
+And here is the equivalent object-based schema:
+
+```js
+// Valid options:
+// "yoda": "warn"
+// "yoda": ["error"]
+// "yoda": ["error", "always"]
+// "yoda": ["error", "never", { "exceptRange": true }]
+// Invalid options:
+// "yoda": ["warn", "never", { "exceptRange": true }, 5]
+// "yoda": ["error", { "exceptRange": true }, "never"]
+module.exports = {
+    meta: {
+        schema: {
+            type: "array",
+            minItems: 0,
+            maxItems: 2,
+            items: [
+                {
+                    enum: ["always", "never"]
+                },
+                {
+                    type: "object",
+                    properties: {
+                        exceptRange: { type: "boolean" }
+                    },
+                    additionalProperties: false
+                }
+            ]
+        }
+    }
+};
+```
+
+Object schemas are more powerful though, and can be much more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specifed (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.
+
+```js
+// Valid options:
+// "yoda": ["error", 6]
+// "yoda": ["error", 5, "strict"]
+// "yoda": ["warn", 10, { someNonOptionalProperty: true }]
+// Invalid options:
+// "yoda": "warn"
+// "yoda": ["error"]
+// "yoda": ["warn", 15]
+// "yoda": ["warn", 7, { }]
+// "yoda": ["error", 3, "on"]
+// "yoda": ["warn", 7, { someOtherProperty: 5 }]
+// "yoda": ["warn", 7, { someNonOptionalProperty: false, someOtherProperty: 5 }]
+module.exports = {
+    meta: {
+        schema: {
+            type: "array",
+            minItems: 1, // Can't specify only severity!
+            maxItems: 2,
+            items: [
+                {
+                    type: "number",
+                    minimum: 0,
+                    maximum: 10
+                },
+                {
+                    anyOf: [
+                        {
+                            type: "object",
+                            properties: {
+                                someNonOptionalProperty: { type: "boolean" }
+                            },
+                            required: ["someNonOptionalProperty"],
+                            additionalProperties: false
+                        },
+                        {
+                            enum: ["off", "strict"]
+                        }
+                    ]
+                }
+            ]
+        }
+    }
+}
+```
+
+And note the easy pitfall of specifying a schema that will _always_ fail because it doesn't account for the array being validated:
```

Comment:
```suggestion
Remember, rule options are always an array, so be careful not to specify a schema for a non-array type at the top level. Here's an example schema that will always fail validation:
And note the easy pitfall of specifying a schema that will always fail because it doesn't account for the array being validated:
```

Reply:
I've also added another sentence to really spell out the consequence of doing this wrong:

> If your schema does not specify an array at the top-level, users can _never_ enable your rule, as their configuration will always be invalid when the rule is enabled.